### PR TITLE
Ridvan namebaseplugin within hsd

### PR DESCRIPTION
**Pull request by rozaydin**
This branch contains the Namebase plugin (within hsd not as a separate repo at least for now)
plugin encapsulates most of the http calls (new coin distribution endpoint i dared not move yet)

Major changes listed:

- deleted open cache completely
- removed custom rpc calls we have, and refactored namebase hnswallet to use respective http calls.
- replaced bid and open http calls with original (non-cached) versions
- created a replica of all endpoints we implemented cache on namebase plugin, i am hoping to move on hns wallet to these endpoints in time.

Things i wish to have included
- moving all caches to namebase plugin, i did a couple tries but found it too risky, and decided to go step by step
- I did testing of endpoint

Please note this PR is dependent on `ridvan-remove-unused-hns-calls` which has a seperate PR.

PS: i do have jsdoc errors but they are not introduced by this PR and most related with hsd code itself. I do have my old PR's jsdoc errored out as well i will fix them.

I am expecting not so easy review ...

### CODE REVIEW

File: lib/wallet/http.js [hunk 1]
```diff
@@ -1687,7 +1354,7 @@ class HTTP extends Server {
 
       const options = TransactionOptions.fromValidator(valid);
 
-      const {result, fromCache} = await req.wallet.withBidCache(
+      const {result, fromCache} = await req.wallet.withUpdateCache(
```
Comment (turbomaze) [inline lib/wallet/http.js:1357]: Wow this looks like a bug nice catch

File: lib/wallet/http.js [hunk 2]
```diff
@@ -2157,6 +1727,38 @@ class HTTP extends Server {
...
+   *
+   * @param {*} httpMethod string value of one of
+   * 'post', 'get', 'put', 'patch', 'delete'
+   * @param {*} path location handler is listening
+   * @param {*} handler http handler that will handle requests
+   * arriving to `path` registered
+   * @param {Object} context context
+   * @returns {undefined}
+   */
+   registerRoute(httpMethod, path, handler, context) {
```
Comment (turbomaze) [inline lib/wallet/http.js:1746]: Nice, would be good to have some way of detecting duplicate paths to avoid bugs but not blocking

File: package-lock.json [hunk 3]
```diff
@@ -1,1409 +0,0 @@
-{
```
Comment (turbomaze) [inline package-lock.json:1]: Why was this file deleted?

File: lib/namebase/index.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  extendWallet(walletPlugin) {
+    this.logger.info('registering http routes');
+    const walletHttp = walletPlugin.http;
+    // set context
+    const context = {
+      walletDB: this.walletDB
+    };
+
+    // Endpoints
+    walletHttp.registerRoute('post', '/wallet/:id/openwithcache',
```
Comment (turbomaze) [inline lib/namebase/index.js:50]: This is new, what is the migration plan? How do we deploy this big change without downtime?

File: lib/namebase/http.js [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+      }
+    });
+
+    return res.json(200, { processedReveals, errors: errorMessages });
+  },
+
+  async handleBatchBid(req, res, context) {
+    // Load Helper classes from context
+    const {
+      TransactionOptions, Validator,network
```
Comment (turbomaze) [inline lib/namebase/http.js:135]: I have an idea to help us deal with all our annoying styling headaches: since it seems difficult to enforce eslint in upstream, we can have an eslint file scoped to this namebase folder and otherwise keep all files in our hsd fork the same. Our styling rules can be as aggressive as we want, based off the airbnb rules etc etc. What do you think?

I want to catch errors like this comma-space thing on this line. I frequently notice style errors in PRs that I try very hard not to comment on, would be very nice to permanently solve that problem

**Review by turbomaze — COMMENTED** [review]
Thanks Ridvan, some questions:

1) What is the migration plan? How do we deploy this without downtime?
2) What do you think of my linting suggestion?
3) Are there any limitations you foresee to extending the wallet HTTP interface via a plugin? Are there going to be events or helpers we want to access but won't be able to?

File: lib/namebase/http.js (hunk 5 above)
Comment (rozaydin) [reply lib/namebase/http.js:135]: Great idea, working on that

**Review by RevCBH — APPROVED** [review]
This all looks reasonable
